aten/src/ATen/core/SingletonSymNodeImpl.h

```diff
@@ -34,12 +35,40 @@ namespace c10 {
 // During tracing the strides of the outputs need to be a function of the size
 // and strides of the inputs so it is important that SingletonSymNode itself is
 // able to express this.
+//
+// NOTE [ SingletonVariant ]
+//
+// Currently, if SingletonSymNodeType::CPP is passed, that means that the
+// singleton is only meant to be used to ferry nested_tensor_size metadata
+// from forward to use in backward. In this case we set `val`, `coeff` etc
+// to bogus values and make sure to error if they are accessed.
+enum class SingletonVariant { PYTHON, CPP };
+
+constexpr c10::DispatchKeySet py_singleton_ks({c10::DispatchKey::Python, c10::DispatchKey::PythonTLSSnapshot});
+constexpr c10::DispatchKeySet cpp_singleton_ks({c10::DispatchKey::NestedTensor});
+
 class TORCH_API SingletonSymNodeImpl : public SymNodeImpl {
  public:
   // CAUTION: you should probably not be constructing these directly; please
-  // the higher-level API in python instead (TODO: actually introduce that).
-  explicit SingletonSymNodeImpl(int64_t val, int64_t coeff)
-      : val_(val), coeff_(coeff) {}
+  // the higher-level API in python instead.
+  explicit SingletonSymNodeImpl(
+      int64_t val,
```

Comment:
val is no longer necessary now that vec is directly stored

Reply:
It is not possible in eq to simply check whether the two vecs are the same because each nested int may be associated multiple vec, e.g. two NestedTensor with same raggedness but on different devices would have two different offsets tensors on different devices, but we'd still want their nested ints to compare equal, and we would do that by mapping both vec to the same val. 

Added this as a comment as well.